src/main/java/com/minecolonies/coremod/util/InventoryUtils.java

```diff
@@ -294,603 +363,1316 @@ public static boolean hasToolLevel(final String tool, final ItemStack stack, fin
+        }
+    }
+
+    /**
+     * Returns a slot number if an {@link IItemHandler} contains given tool
+     * type.
+     *
+     * @param itemHandler the {@link IItemHandler} to get the slot from.
+     * @param tool        the tool type to look for.
+     * @return slot number if found, -1 if not found.
+     */
+    public static int getFirstSlotOfItemHandlerContainingTool(@NotNull final IItemHandler itemHandler, @NotNull final String tool)
+    {
+        return findFirstSlotInItemHandlerWith(itemHandler,
+          (ItemStack stack) -> (!isItemStackEmpty(stack) && (stack.getItem().getToolClasses(stack).contains(tool) || ("hoe".equals(tool) && stack.getUnlocalizedName()
+                                                                                                                                              .contains("hoe"))
+                                                               || ("rod".equals(tool) && stack.getUnlocalizedName().contains("fishingRod")))));
+    }
 
-            if (slot < 0)
+    /**
+     * Verifies if there is one tool with an acceptable level
+     * in a worker's inventory.
+     *
+     * @param toolTypeName  the type of tool needed
+     * @param itemHandler   the worker's inventory
+     * @param requiredLevel the worker's hut level
+     * @return true if tool is acceptable
+     */
+    public static boolean hasItemHandlerToolWithLevel(@NotNull final IItemHandler itemHandler, final String toolTypeName, final int requiredLevel)
+    {
+        return findFirstSlotInItemHandlerWith(itemHandler,
+          (ItemStack stack) -> (!isItemStackEmpty(stack) && (Utils.isTool(stack, toolTypeName) && verifyToolLevel(stack,
+            Utils.getMiningLevel(stack, toolTypeName),
+            requiredLevel)))) > -1;
+    }
+
+    /**
+     * Method to swap the ItemStacks from the given source {@link IItemHandler}
+     * to the given target {@link ICapabilityProvider}.
+     *
+     * @param sourceHandler  The {@link IItemHandler} that works as Source.
+     * @param sourceIndex    The index of the slot that is being extracted
+     *                       from.
+     * @param targetProvider The {@link ICapabilityProvider} that works as
+     *                       Target.
+     * @return True when the swap was successful, false when not.
+     */
+    public static boolean transferItemStackIntoNextFreeSlotInProvider(
+                                                                       @NotNull final IItemHandler sourceHandler,
+                                                                       @NotNull int sourceIndex,
+                                                                       @NotNull ICapabilityProvider targetProvider)
+    {
+        for (IItemHandler handler : getItemHandlersFromProvider(targetProvider))
+        {
+            if (transferItemStackIntoNextFreeSlotInItemHandlers(sourceHandler, sourceIndex, handler))
      
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Introduce a new variable instead of reusing the parameter "itemStack". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1226)